### ISSUE TITLE
Matching on structs defined from another module takes significantly longer time to type check

### ISSUE THREAD

**Issue by philnguyen** [racket/typed-racket#553]
### What version of Racket are you using?
`v6.9.0.4`

### What program did you run?
file `define.rkt`:
```racket
#lang typed/racket/base

(provide (all-defined-out))

(struct s1  ([x : Any]) #:transparent)
(struct s2  ([x : Any]) #:transparent)
(struct s3  ([x : Any]) #:transparent)
(struct s4  ([x : Any]) #:transparent)
(struct s5  ([x : Any]) #:transparent)
(struct s6  ([x : Any]) #:transparent)
(struct s7  ([x : Any]) #:transparent)
(struct s8  ([x : Any]) #:transparent)
(struct s9  ([x : Any]) #:transparent)
(struct s10 ([x : Any]) #:transparent)
(struct s11 ([x : Any]) #:transparent)
(struct s12 ([x : Any]) #:transparent)
```
file `main.rkt`:
```racket
#lang typed/racket/base

(require racket/match
         "define.rkt")

(define f
  (match-lambda
    [(or (s1  x)
         (s2  x)
         (s3  x)
         (s4  x)
         (s5  x)
         (s6  x)
         (s7  x)
         (s8  x)
         (s9  x)
         (s10 x)
         (s11 x)
         (s12 x))
     42]))
```

### What should have happened?
Should type check about as fast as the below combined module.
Unintuitively, the combined module type checks immediately, while for the split modules,
`main.rkt` takes 1+ minute to finish.

file `combined.rkt`:
```racket
#lang typed/racket/base

(require racket/match)

(struct s1  ([x : Any]) #:transparent)
(struct s2  ([x : Any]) #:transparent)
(struct s3  ([x : Any]) #:transparent)
(struct s4  ([x : Any]) #:transparent)
(struct s5  ([x : Any]) #:transparent)
(struct s6  ([x : Any]) #:transparent)
(struct s7  ([x : Any]) #:transparent)
(struct s8  ([x : Any]) #:transparent)
(struct s9  ([x : Any]) #:transparent)
(struct s10 ([x : Any]) #:transparent)
(struct s11 ([x : Any]) #:transparent)
(struct s12 ([x : Any]) #:transparent)

(define f
  (match-lambda
    [(or (s1  x)
         (s2  x)
         (s3  x)
         (s4  x)
         (s5  x)
         (s6  x)
         (s7  x)
         (s8  x)
         (s9  x)
         (s10 x)
         (s11 x)
         (s12 x))
     42]))
```

**Comment by pnwamk** [comment]
__TL;DR__ it looks like TR's exporting affects how `match` expands w.r.t. structs in a way that negatively impacts type checking =(  ...perhaps @samth (who is a SME on both) has some useful insights. 

__Side note__: a good rule of thumb in TR when using `match` is to avoid `or` patterns. By not using an `or` and adding a type annotation to the function, type checking time in either case (i.e. with local or non-local struct definitions) is more or less equivalent.


__Long version__, I've done at least some preliminary digging on this, and here's what I've got so far:

One difference that is at least part of the reason why this is so slow is because TR changes the way struct identifiers are defined/exported, which in turn changes how `match` is able to optimize things before we typecheck. (_as an aside, I'm making this claim primarily from observing the expansion of match and noting how complicated exports are for TR, not because I've taken the time yet to fully understand all the tweaks involved in our identifier export process in TR_)

In regular Racket, the `match` macro will convert the match clause `[(s1 x) . rhs]` into something like

```
(if (s1? arg)
     (let ([x (unsafe-struct-ref arg 0)]) . rhs)
      ...)
```

but in TR today, when the struct `s1` is defined in another module this does not happen. The `match` clauses are expanded and we instead end up type checking the following:

```
(if (s1? arg)
     (let ([x (s1-x arg)]) . rhs)
      ...)
```

and the expression `(s1-x arg)` ends up having more logical "baggage" associated with it that can result in exponential growth in the size of the logical info the type checker works with.

I think it would be nice if TR's exporting did not affect match in this way.

**Comment by mfelleisen** [comment]
> On May 21, 2017, at 10:35 AM, Andrew Kent <notifications@github.com> wrote:
> 
> TL;DR it looks like TR's exporting affects how match expands w.r.t. structs in a way that negatively impacts type checking =( ...perhaps @samth <https://github.com/samth> (who is a SME on both) has some useful insights.
> 
> Side note: a good rule of thumb in TR when using match is to avoid or patterns. By not using an or and adding a type annotation to the function, type checking time in either case (i.e. with local or non-local struct definitions) is more or less equivalent.
> 
> Long version, I've done at least some preliminary digging on this, and here's what I've got so far:
> 
> One difference that is at least part of the reason why this is so slow is because TR changes the way struct identifiers are defined/exported, which in turn changes how match is able to optimize things before we typecheck. (as an aside, I'm making this claim primarily from observing the expansion of match and noting how complicated exports are for TR, not because I've taken the time yet to fully understand all the tweaks involved in our identifier export process in TR)
> 
> In regular Racket, the match macro will convert the match clause [(s1 x) . rhs] into something like
> 
> (if (s1? arg)
>      (let ([x (unsafe-struct-ref arg 0)]) . rhs)
>       ...)
> but in TR today, when the struct s1 is defined in another module this does not happen. The match clauses are expanded and we instead end up type checking the following:
> 
> (if (s1? arg)
>      (let ([x (s1-x arg)]) . rhs)
>       ...)
> and the expression (s1-x arg) ends up having more logical "baggage" associated with it that can result in exponential growth in the size of the logical info the type checker works with.
> 
> I think it would be nice if TR's exporting did not affect match in this way
> 


Once this last expression type-checks, the TR optimizer could bring back the (unsafe-struct-ref . 0) expression. Indeed, all type-checked struct expressions could benefit from this optimization and I some point I thought Vincent and/or Sam had done something like that.

**Comment by pnwamk** [comment]
I'm sorry---I omitted some details and/or worded things confusingly.

 Indeed, those sorts of optimizations are performed by the TR optimizer (thanks Vincent and/or Sam!). However, those optimizations occur _after_ type checking---I was suggesting that it would be nice if `match` could perform the optimization for TR code the same way it does for regular Racket code in these cases, but I think currently how TR exports code prevents `match` from performing the optimization.

**Comment by mfelleisen** [comment]
If the optimizer transforms the code anyway, why does is the code so slow? 

I can see that TR wouldn’t be able to check the optimized code — at least currently. If we wrote TR as an elaborator from the syntax-propertied surface syntax (kernel) to plain surface syntax (also kernel). it could change expressions for which it can prove types. For example, (unsafe-struct-ref . 0) could be introduced as part of type checking instead of waiting for the optimizer to kick in. Indeed, we might then be able to build in the entire trivial package into TR and even provide an API for adding rewriting rules.

**Comment by samth** [comment]
@mfelleisen TR type checks the optimized code just fine. The speed issue is all in type checking time, not run time.

**Comment by samth** [comment]
@pnwamk The difference is that `match` knows that struct bindings are really provided by the `struct` macro when they are. The ones provided by TR wrapping can't have that authenticity (because they aren't authentic). Maybe we can improve the `provide` code to avoid that problem, but it seems like the bigger issue is the slower type checking with regular accessors.

**Comment by mfelleisen** [comment]
> On May 22, 2017, at 9:17 AM, Sam Tobin-Hochstadt <notifications@github.com> wrote:
> 
> @pnwamk <https://github.com/pnwamk> The difference is that match knows that struct bindings are really provided by the struct macro when they are. The ones provided by TR wrapping can't have that authenticity (because they aren't authentic). Maybe we can improve the provide code to avoid that problem, but it seems like the bigger issue is the slower type checking with regular accessors.
> 

If TR type-checks the unsafe code just fine, why do you export regular accessors for this case?

**Comment by samth** [comment]
TR exports the same accessors in both cases. The difference is, as I mentioned in my comment to @pnwamk, that `match` can tell that the struct bindings are really from `struct` and `match decides to use the `unsafe-struct-ref` accessors.

**Comment by mfelleisen** [comment]
Thanks for clarifying.
